host/ecallids.c

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include "ecallids.h"
+#include <openenclave/internal/raise.h>
+#include "hostthread.h"
+
+/* The maximum numer of different ecalls that an application can make is
+ * bounded. */
+static const char* _ecall_table[OE_MAX_ECALLS];
+static uint32_t _ecall_table_size;
+
+/* Mutex for assigning/looking up global ids in a thread-safe manner. */
+static oe_mutex _lock;
+
+/* Get the global ecall id from the _ecall_table. Locking must be done by
+ * caller. */
+static oe_result_t _get_global_id(const char* name, uint64_t* global_id)
+{
+    oe_result_t result = OE_NOT_FOUND;
+    if (!name || !global_id)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Search for id assigned for given name. */
+    for (uint64_t i = 0; i < _ecall_table_size; i++)
+    {
+        if (strcmp(_ecall_table[i], name) == 0)
+        {
+            *global_id = i;
+            break;
+        }
+    }
+
+    /* If the name is not found, adding it to the table. */
+    if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
+    {
+        if (_ecall_table_size >= OE_MAX_ECALLS)
+        {
+            *global_id = OE_MAX_ECALLS;
+            OE_RAISE(OE_OUT_OF_BOUNDS);
+        }
+        _ecall_table[_ecall_table_size] = name;
+        *global_id = _ecall_table_size;
+        _ecall_table_size++;
+    }
+
+    result = OE_OK;
+done:
+    return result;
+}
+
+oe_result_t oe_get_global_id(const char* name, uint64_t* global_id)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    bool locked = false;
+    if (oe_mutex_lock(&_lock) != 0)
+        OE_RAISE(OE_FAILURE);
+    locked = true;
+
+    result = _get_global_id(name, global_id);
+done:
+    if (locked)
+        oe_mutex_unlock(&_lock);
+    return result;
+}
+
+oe_result_t oe_get_ecall_ids(
+    oe_enclave_t* enclave,
+    const char* name,
+    uint64_t* global_id,
+    uint64_t* id)
+{
+    oe_result_t result = OE_FAILURE;
+    oe_ecall_id_t* ecall_id_table = NULL;
+
+    /* Validate parameters */
+    if (!enclave || !global_id || !name || !id)
+    {
+        result = OE_INVALID_PARAMETER;
+        goto done;
+    }
+
+    /* Initialize output parameter to NULL id */
+    *id = OE_ECALL_ID_NULL;
+
+    /* Fetch ecall_id_table using platform agnostic API */
+    OE_CHECK(oe_get_ecall_id_table(enclave, &ecall_id_table));
+
+    /* Lookup/assign global id */
+    if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
+    {
+        /* The global_id variable is generated by edger8r, one global id
+         *  variable per ecall. The same variable will be reused every time that
+         * ecall is made. Therefore, for an ecall, the following code will be
+         * executed only once.
+         */
+        OE_CHECK(oe_get_global_id(name, global_id));
+    }
+
+    /* Look up function id based on global id. */
+    if (*global_id >= OE_MAX_ECALLS)
```

Comment:
per my earlier comment, seems safer to return and check against OE_GLOBAL_ECALL_ID_NULL here

Reply:
In the latest version, the check reads better:
`if (*global_id >= ecall_id_table_size)`